File: cses/05_range_queries/1736_polynomial-queries/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    q = int(data[1])
    pos = 2

    # Prefix sum của mảng ban đầu (tĩnh, không bao giờ thay đổi).
    pre = [0] * (n + 1)
    s = 0
    for i in range(1, n + 1):
        s += int(data[pos]); pos += 1
        pre[i] = s

    # Ba Fenwick tree lưu f[j], j*f[j], j^2*f[j]. Kích thước n là đủ vì các
    # query prefix chỉ đến n; các point update tại r+1, r+2 vượt quá n bị bỏ qua.
    a0 = [0] * (n + 1)
    a1 = [0] * (n + 1)
    a2 = [0] * (n + 1)

    out = []
    ap = out.append

    for _ in range(q):
        t = data[pos]; a = int(data[pos + 1]); b = int(data[pos + 2]); pos += 3
        if t == b'1':
            # Update loại 1: cộng cấp số cộng trên [l, r] bằng ba point update sai
            # phân bậc hai vào cả ba BIT (đã inline vòng lặp BIT cho nhanh).
            l = a; r = b
            # Điểm l: hệ số (1, l, l*l) cho (a0, a1, a2).
            i = l; v1 = l; v2 = l * l
            while i <= n:
                a0[i] += 1; a1[i] += v1; a2[i] += v2
                i += i & (-i)
            # Điểm r+1: hệ số c1 = -(r-l+2).
            p = r + 1
            if p <= n:
                c1 = -(r - l + 2); w1 = p * c1; w2 = p * p * c1
                i = p
                while i <= n:
                    a0[i] += c1; a1[i] += w1; a2[i] += w2
                    i += i & (-i)
            # Điểm r+2: hệ số c2 = (r-l+1).
            p = r + 2
            if p <= n:
                c2 = r - l + 1; w1 = p * c2; w2 = p * p * c2
                i = p
                while i <= n:
                    a0[i] += c2; a1[i] += w1; a2[i] += w2
                    i += i & (-i)
        else:
            # Query loại 2: tổng đoạn [l, r] = (PS(r) - PS(l-1)) + phần ban đầu.
            l = a; r = b
            # PS(r): ba prefix query trên BIT rồi ghép theo đa thức bậc hai.
            x = r; s0 = 0; s1 = 0; s2 = 0; i = x
            while i > 0:
                s0 += a0[i]; s1 += a1[i]; s2 += a2[i]
                i -= i & (-i)
            psr = ((x * x + 3 * x + 2) * s0 - (2 * x + 3) * s1 + s2) // 2
            # PS(l-1): tương tự, bằng 0 nếu l-1 <= 0.
            x = l - 1
            if x > 0:
                s0 = 0; s1 = 0; s2 = 0; i = x
                while i > 0:
                    s0 += a0[i]; s1 += a1[i]; s2 += a2[i]
                    i -= i & (-i)
                psl = ((x * x + 3 * x + 2) * s0 - (2 * x + 3) * s1 + s2) // 2
            else:
                psl = 0
            ap(psr - psl + pre[r] - pre[l - 1])

    sys.stdout.buffer.write(('\n'.join(map(str, out)) + '\n').encode())


main()

File: cses/05_range_queries/1736_polynomial-queries/solution.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <string>
#include <vector>

using i64 = long long;
using i128 = __int128_t;

// Fenwick tree (BIT): point update cộng dồn, prefix query lấy tổng tiền tố.
class Fenwick {
public:
    explicit Fenwick(int n) : data_(static_cast<std::size_t>(n + 1), 0) {}

    void add(int position, i128 value) {
        const int n = static_cast<int>(data_.size()) - 1;
        for (int i = position; i <= n; i += i & -i) {
            data_[static_cast<std::size_t>(i)] += value;
        }
    }

    [[nodiscard]] i128 prefix_sum(int position) const {
        i128 result = 0;
        for (int i = position; i > 0; i -= i & -i) {
            result += data_[static_cast<std::size_t>(i)];
        }
        return result;
    }

private:
    std::vector<i128> data_;
};

// In số kiểu i128 (std::cout không hỗ trợ trực tiếp kiểu này).
void print_integer(i128 value) {
    if (value == 0) {
        std::cout << '0';
        return;
    }
    if (value < 0) {
        std::cout << '-';
        value = -value;
    }
    std::string digits;
    while (value > 0) {
        digits.push_back(static_cast<char>('0' + static_cast<int>(value % 10)));
        value /= 10;
    }
    std::reverse(digits.begin(), digits.end());
    std::cout << digits;
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n = 0;
    int q = 0;
    std::cin >> n >> q;

    // Prefix sum tĩnh của mảng ban đầu (không bao giờ thay đổi).
    std::vector<i64> initial_prefix(static_cast<std::size_t>(n + 1), 0);
    for (int i = 1; i <= n; ++i) {
        i64 value = 0;
        std::cin >> value;
        initial_prefix[static_cast<std::size_t>(i)] =
            initial_prefix[static_cast<std::size_t>(i - 1)] + value;
    }

    // Ba BIT lưu prefix sum của f[j], j*f[j], j^2*f[j] (tức A0, A1, A2).
    Fenwick moment0(n);
    Fenwick moment1(n);
    Fenwick moment2(n);

    // Cộng một điểm sai phân bậc hai vào cả ba BIT; vị trí vượt quá n thì bỏ qua.
    const auto add_second_difference = [&](int position, i128 coefficient) {
        if (position > n) {
            return;
        }
        const i128 p = position;
        moment0.add(position, coefficient);
        moment1.add(position, p * coefficient);
        moment2.add(position, p * p * coefficient);
    };

    // PS(x): prefix sum của phần cộng thêm, tính theo đa thức bậc hai của x.
    const auto added_prefix_sum = [&](int x) -> i128 {
        if (x <= 0) {
            return 0;
        }
        const i128 xx = x;
        const i128 a0 = moment0.prefix_sum(x);
        const i128 a1 = moment1.prefix_sum(x);
        const i128 a2 = moment2.prefix_sum(x);
        return ((xx * xx + 3 * xx + 2) * a0 - (2 * xx + 3) * a1 + a2) / 2;
    };

    while (q-- > 0) {
        int type = 0;
        int left = 0;
        int right = 0;
        std::cin >> type >> left >> right;
        if (type == 1) {
            // Update loại 1: cấp số cộng trên [left, right] rút về ba point update.
            add_second_difference(left, 1);
            add_second_difference(right + 1, -static_cast<i128>(right - left + 2));
            add_second_difference(right + 2, static_cast<i128>(right - left + 1));
        } else {
            // Query loại 2: tổng đoạn = phần ban đầu cộng phần cộng thêm.
            const i128 original =
                initial_prefix[static_cast<std::size_t>(right)] -
                initial_prefix[static_cast<std::size_t>(left - 1)];
            const i128 answer = original + added_prefix_sum(right) - added_prefix_sum(left - 1);
            print_integer(answer);
            std::cout << '\n';
        }
    }
}
